compiler/rustc_mir_transform/src/unreachable_prop.rs

```diff
@@ -57,42 +68,40 @@ impl MirPass<'_> for UnreachablePropagation {
 
 fn remove_successors<'tcx, F>(
     terminator_kind: &TerminatorKind<'tcx>,
-    predicate: F,
+    is_unreachable: F,
 ) -> Option<TerminatorKind<'tcx>>
 where
     F: Fn(BasicBlock) -> bool,
 {
-    let terminator = match *terminator_kind {
-        TerminatorKind::Goto { target } if predicate(target) => TerminatorKind::Unreachable,
-        TerminatorKind::SwitchInt { ref discr, switch_ty, ref targets } => {
+    let terminator = match terminator_kind {
+        // This will unconditionally run into an unreachable and is therefore unreachable as well.
+        TerminatorKind::Goto { target } if is_unreachable(*target) => TerminatorKind::Unreachable,
+        TerminatorKind::SwitchInt { targets, .. } => {
             let otherwise = targets.otherwise();
 
-            let original_targets_len = targets.iter().len() + 1;
-            let (mut values, mut targets): (Vec<_>, Vec<_>) =
-                targets.iter().filter(|(_, bb)| !predicate(*bb)).unzip();
+            let otherwise_is_unreachable = is_unreachable(otherwise);
 
-            if !predicate(otherwise) {
-                targets.push(otherwise);
-            } else {
-                values.pop();
-            }
-
-            let retained_targets_len = targets.len();
-
-            if targets.is_empty() {
+            // This case is equivalent to a goto terminator.
+            if targets.all_targets().len() == 0 && otherwise_is_unreachable
+            // If all targets are unreachable, we can be unreachable as well.
+                || otherwise_is_unreachable && targets.iter().all(|(_, bb)| is_unreachable(bb))
```

Comment:
```suggestion
if targets.all_targets().iter().all(|bb| is_unreachable(*bb))
```

`otherwise` is one of the `all_targets`